File: torrent-client/debrid_services/realdebrid.py
import requests

class RealDebrid:
    def __init__(self, api_key):
        self.api_key = api_key
        self.base_url = "https://api.real-debrid.com/rest/1.0/"

    def get_user_info(self):
        url = self.base_url + "user"
        headers = {
            "Authorization": "Bearer " + self.api_key
        }
        response = requests.get(url, headers=headers)
        return response.json()

    def add_magnet_link(self, magnet):
        url = self.base_url + "torrents/addMagnet"
        headers = {
            "Authorization": "Bearer " + self.api_key
        }
        data = {
            "magnet": magnet
        }
        response = requests.post(url, headers=headers, data=data)
        return response.json()

    def select_all_files(self, torrent_id):
        url = self.base_url + "torrents/selectFiles/" + torrent_id
        headers = {
            "Authorization": "Bearer " + self.api_key
        }
        data = {
            "files": "all"
        }
        response = requests.post(url, headers=headers, data=data)
        return response

    def get_torrent_info(self, torrent_id):
        url = self.base_url + "torrents/info/" + torrent_id
        headers = {
            "Authorization": "Bearer " + self.api_key
        }
        response = requests.get(url, headers=headers)
        return response.json()

    def get_direct_link(self, link):
        url = self.base_url + "unrestrict/link"
        headers = {
            "Authorization": "Bearer " + self.api_key
        }
        data = {
            "link": link
        }
        response = requests.post(url, headers=headers, data=data)
        return response.json()

File: src/main/services/real-debrid.ts
import axios, { AxiosInstance } from "axios";
import parseTorrent from "parse-torrent";
import type {
  RealDebridAddMagnet,
  RealDebridTorrentInfo,
  RealDebridUnrestrictLink,
  RealDebridUser,
} from "@types";

export class RealDebridClient {
  private static instance: AxiosInstance;
  private static baseURL = "https://api.real-debrid.com/rest/1.0";

  static authorize(apiToken: string) {
    this.instance = axios.create({
      baseURL: this.baseURL,
      headers: {
        Authorization: `Bearer ${apiToken}`,
      },
    });
  }

  static async addMagnet(magnet: string) {
    const searchParams = new URLSearchParams({ magnet });

    const response = await this.instance.post<RealDebridAddMagnet>(
      "/torrents/addMagnet",
      searchParams.toString()
    );

    return response.data;
  }

  static async getTorrentInfo(id: string) {
    const response = await this.instance.get<RealDebridTorrentInfo>(
      `/torrents/info/${id}`
    );
    return response.data;
  }

  static async getUser() {
    const response = await this.instance.get<RealDebridUser>(`/user`);
    return response.data;
  }

  static async selectAllFiles(id: string) {
    const searchParams = new URLSearchParams({ files: "all" });

    await this.instance.post(
      `/torrents/selectFiles/${id}`,
      searchParams.toString()
    );
  }

  static async unrestrictLink(link: string) {
    const searchParams = new URLSearchParams({ link });

    const response = await this.instance.post<RealDebridUnrestrictLink>(
      "/unrestrict/link",
      searchParams.toString()
    );

    return response.data;
  }

  private static async getAllTorrentsFromUser() {
    const response =
      await this.instance.get<RealDebridTorrentInfo[]>("/torrents");

    return response.data;
  }

  static async getTorrentId(magnetUri: string) {
    const userTorrents = await RealDebridClient.getAllTorrentsFromUser();

    const { infoHash } = await parseTorrent(magnetUri);
    const userTorrent = userTorrents.find(
      (userTorrent) => userTorrent.hash === infoHash
    );

    if (userTorrent) return userTorrent.id;

    const torrent = await RealDebridClient.addMagnet(magnetUri);
    return torrent.id;
  }
}
